proposals/0023-scalar-constant-buffer-layout.md

```diff
@@ -0,0 +1,300 @@
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer ...
+
+[bufferlayout("legacy")]
+ConstantBuffer<MyStruct> MyBuffer ...
+
+[bufferlayout("scalar")]
+ConstantBuffer<MyStruct> MyBuffer ...
+```
+
+The following example demonstrates the difference in buffer alignment and size
+between legacy and scalar layouts in DXIL.
+Note that in legacy layout the float3 is aligned such that it begins in the
+second 16 byte space, while in scalar layout it begins immediately after the
+float2, with the 3rd element starting at offset 16.
+
+```
+[bufferlayout("legacy")]
+cbuffer MyBuffer {
+  float2 Foo;
+  float3 Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float2 Foo;                                   ; Offset:    0
+;       float3 Bar;                                   ; Offset:   16
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    28
+;
+; }
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer {
+  float2 Foo;
+  float3 Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float2 Foo;                                   ; Offset:    0
+;       float3 Bar;                                   ; Offset:    8
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    20
+;
+; }
+```
+
+As all values must be aligned to their scalar size, a scalar or element of a
+composite type may not lay across the 16 byte boundary.
+In the following example the double is aligned identically in both legacy and
+scalar layout, to 8 bytes, demonstrating this point.
+
+```
+[bufferlayout("legacy")]
+cbuffer MyBuffer {
+  float3 Foo;
+  double Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float3 Foo;                                   ; Offset:    0
+;       double Bar;                                   ; Offset:   16
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    24
+;
+; }
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer {
+  float3 Foo;
+  double Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float3 Foo;                                   ; Offset:    0
+;       double Bar;                                   ; Offset:   16
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    24
+;
+; }
+```
+
+Matrix types in legacy layout are significantly different to scalar layout.
+With legacy layout a matrix is laid out so each row is 16 byte aligned.
+The exception to this is a matrix of doubles, in which case a double3x and
+double4x matrix will have the rows aligned to 32 bytes.
+With scalar layouts, as with vectors, each element is aligned to it's own scalar
+size.
+The following example demonstrates the differences in matrix layout by
+showing the cbuffer loads and extractvalue instructions needed to load each
+element of a 2x2 float matrix.
+Note that the legacy example requires two 16 byte loads from the cbuffer, while
+the scalar example requires only one.
```

Comment:
As someone not entirely familiar with DXIL, it would be useful to explain here _why_ the legacy version needs an extra load, because it's not at all clear here.

Reply:
The cbuffer storage for a matrix<T,2,2> is basically a vector<T,2>[2]. The packing rules for arrays in cbuffers are that each array index starts on a new "row" of the cbuffer. You can see this visualized here:
https://maraneshi.github.io/HLSL-ConstantBufferLayoutVisualizer/?visualizer=MYIwrgZhCmBOAE0AeBDAtgBwDbXgbwCh5j4IsB7FAFwCYkb41qBuAgX2aA

Reply:
Ah thanks, that does help quite a bit - would be good to get this explanation added to the proposal!